jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -581,14 +581,19 @@ private Pool.Entry<RetainableByteBuffer> evict()
         }
     }
 
-    private static class Buffer extends AbstractRetainableByteBuffer
+    private static class PooledBuffer extends RetainableByteBuffer.Pooled
```

Comment:
The reason it exists is that it allows a Pool reference to be carried with a buffer so it can be used if extra allocations are needed.   Some methods in RBB also use the knowledge that the buffer is pool in their heuristic considerations

Reply:
It could almost be an internal class, but then it would spread the logic around too much..

Reply:
A Pooled buffer is one that knows which pool it came from, so that `copy` operations can take buffers from the same pool and `slice` operations can preserve the pool reference.   So there is more behaviour than just the optimization in `takeByteArray`.   We have a public interface for `ByteBufferPool` and a public interface for `RetainableByteBuffer`, so I don't see a problem with a public API representing a `RetainableByteBuffer` from a `ByteBufferPool`.